### PR TITLE
Updated and generalized more promotions

### PR DESCRIPTION
**Pull request by xlenstra**
This PR is the follow-up for #4292, generalizing and updating basically all other promotions.
Again, each commit only changes a single promotion to allow for easier reviewing.
The following promotion was added:
- Supply: Water units can heal outside of friendly territory (15 hp/turn). Note that in the base game this was locked behind Tier II promotions. I choose to lock it behind Tier III promotions instead as it is very powerful in my opinion.

Added uniques:
- "[amount] extra interceptions may be made per turn"
- "[signedAmount]% Strength when defending vs [unitFilter]"
- "[signedAmount]% Strength for enemy [unitFilter] units in adjacent tiles"
- "May heal outside of friendly territory" - for water units
- [signedAmount]% Damage when intercepting

Deprecated uniques:
- "1 extra Interception may be made per turn"
- "+25% Defence against ranged attacks"
- "-10% combat strength for adjacent enemy units"
- "Bonus when intercepting [amount]%"

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
> Cover+33, Ambush+33

You sure? [This](https://civilization.fandom.com/wiki/Promotions_(Civ5)) doesn't mention somesuch

> I choose to lock it behind Tier III

I'd leave such intentional "deviations" from original rules to mods.
Also, I would interpret the wiki as 15 _total_ per turn. Wouldn't these uniques add up two healing sources? (just a doubt, I didn't check)

**Comment by xlenstra** [comment]
> > Cover+33, Ambush+33
> 
> You sure? [This](https://civilization.fandom.com/wiki/Promotions_(Civ5)) doesn't mention somesuch
> 

I checked the ingame civilopedia with G&K enabled and BNW disabled, I don't think it can get more accurate than that

> > I choose to lock it behind Tier III
> 
> I'd leave such intentional "deviations" from original rules to mods.

I'll leave this to Yairm to decide. It is really strong and basically removes one of the major downsides of a navy.

> Also, I would interpret the wiki as 15 _total_ per turn. Wouldn't these uniques add up two healing sources? (just a doubt, I didn't check)

The way it is implemented now, is that water units still heal 0 per turn when outside friendly territory, but with this unique they are allowed to have other uniques providing them bonuses. This is so that mod makers can choose themselves how strong the Supply bonus is. Relveant code is in in the `rankTileForHealing()` in the `mapUnit` class.

**Comment by SomeTroglodyte** [comment]
:+1: Oooooooooooh so you have access to      _t h e   O r i g i n a l_     ...
OK, you're the official new primary source of accurate information :wink: ... Besides - thorough work - kudos.

File: core/src/com/unciv/logic/battle/BattleDamage.kt [hunk 1]
```diff
@@ -51,17 +51,19 @@ object BattleDamage {
...
+                    && adjacentUnits.any { it.civInfo == civInfo && it.matchesFilter(unique.params[2]) } 
                 ) {
-                    adjacentUnitBonus += unique.params[0].toInt()
+                    modifiers.add("Adjacent units", unique.params[0].toInt())
                 }
             }
-            if (adjacentUnitBonus != 0)
-                modifiers["Adjacent unit"] = adjacentUnitBonus
+            
+            for (unique in adjacentUnits.flatMap { it.getMatchingUniques("[]% Strength for [] units in adjacent tiles you are at war with") })
```
Comment (yairm210) [inline core/src/com/unciv/logic/battle/BattleDamage.kt:64]: Do you think 'you are at war with' is more clear-cut than 'enemy'?

File: core/src/com/unciv/logic/map/MapUnit.kt [hunk 2]
```diff
@@ -842,9 +845,13 @@ class MapUnit {
     }
 
     fun canIntercept(attackedTile: TileInfo): Boolean {
-        if (attacksThisTurn > 1) return false
         if (interceptChance() == 0) return false
-        if (attacksThisTurn > 0 && !hasUnique("1 extra Interception may be made per turn")) return false
+        val maxAttacksPerTurn = 1 + 
+            getMatchingUniques("[] extra interceptions may be made per turn").sumBy { it.params[0].toInt() } + 
+            // Deprecated since 3.15.7
+                getMatchingUniques("1 extra interception may be made per turn").sumBy { 1 }
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/MapUnit.kt:852]: sumBy { 1 } -> .count()

File: core/src/com/unciv/logic/battle/BattleDamage.kt (hunk 1 above)
Comment (xlenstra) [reply core/src/com/unciv/logic/battle/BattleDamage.kt:64]: The main problem I found when doing these kind of things, is that the order of adjectives doesn't line up.
For example when using "[]% Strength for enemy [] units" you also have "[]% Strength for enemy All units" which doesn't sound right, but if you use "[]% Strength for [] enemy units" you also have "[]% Strength for land enemy units" which also doesn't sound right. Moving the enemy to a different spot solves this, but does make it longer.

For now I've changed it to "[]% Strength for all enemy [] units in adjacent [tileFilter] tiles", which has the same problem but with the tileFilter, but that is the way it is.

Comment (yairm210) [reply core/src/com/unciv/logic/battle/BattleDamage.kt]: Sounds thought-out :)

**Pull request merged**